### PR TITLE
Tests for frontier manager

### PR DESCRIPTION
**Pull request by Preetwinder**
Fake backend, canonicalsolver and middlewares are used. There are a few different middlewares, some don't pass the values through, some modify some values etc. Each of these component records all the requests, responses, errors in lists. In the tests we verify the type and content of values passed through, and whether they were modified.

### CODE REVIEW

File: frontera/tests/mocks/components.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        for i in range(max_next_requests):
+            if self.requests:
+                lst.append(self.requests.pop())
+        return lst
+
+    def request_error(self, request, error):
+        self.errors.append((request, error))
+
+
+class FakeMiddleware(Middleware):
```
Comment (sibiryakov) [inline frontera/tests/mocks/components.py]: Is it possible to unify FakeMiddleware and FakeBackend? They look quite similar.

File: frontera/tests/test_frontier_manager.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        assert set([link for link in fm.backend.links]) == set([r2, r3])
+        assert set([link for link in fm.canonicalsolver.links]) == set([r2, r3])
+        assert [set([link for link in mw.links]) for mw in fm.middlewares] == [set([r2, r3])]*4
+        assert [link.meta['test_links'] for link in [r2, r3]] == ['test']*2
+        assert [link.meta['test_links_canonical_solver'] for link in [r2, r3]] == ['test']*2
+
+    def test_get_next_requests(self):
+        fm = self.setup_frontier_manager()
+        fm.backend.put_requests([r1, r2, r3])
+        assert set(fm.get_next_requests(3)) == set([r1, r2, r3])
```
Comment (sibiryakov) [inline frontera/tests/test_frontier_manager.py]: what if we add here asserts for iteration and n_requests?

**Comment by sibiryakov** [comment]
Let's add test for finishing condition implemented in Backend also. 
In general, it looks great! I like it.

File: frontera/tests/mocks/components.py (hunk 1 above)
Comment (Preetwinder) [inline frontera/tests/mocks/components.py]: I have made the FakeBackend a subclass of the FakeMiddleware.

**Comment by Preetwinder** [comment]
I have added a test for max_requests reached, and also added checks for iteration and n_requests.

**Comment by sibiryakov** [comment]
Looks great!

**Pull request merged**